src/combinations.rs

```diff
@@ -0,0 +1,126 @@
+use std::iter::Iterator;
+use std::mem::MaybeUninit;
+
+/// An iterator that returns k-length combinations of values from `iter`.
+///
+/// This `struct` is created by the [`combinations`] method on [`IterExt`]. See its
+/// documentation for more.
+///
+/// [`combinations`]: super::IterExt::combinations
+/// [`IterExt`]: super::IterExt
+#[derive(Clone, Debug)]
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+pub struct Combinations<I, const K: usize>
+where
+    I: Iterator,
+    I::Item: Clone,
+{
+    iter: I,
+    buffer: Vec<I::Item>,
+    indices: [usize; K],
+    first: bool,
+}
+
+impl<I, const K: usize> Combinations<I, K>
+where
+    I: Iterator,
+    I::Item: Clone,
+{
+    // NOTE: this clippy attribute can be removed once we can `collect` into `[usize; K]`.
+    #[allow(clippy::clippy::needless_range_loop)]
+    pub(crate) fn new(mut iter: I) -> Self {
+        // Prepare the indices.
+        let mut indices = [0; K];
+        for i in 0..K {
```

Comment:
I would move the clippy arrtibute and the comment right above the loop here, the scope seems too broad now.